sql/limit.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// limit constructs a limitNode based on the LIMIT and OFFSET clauses.
+func (*planner) limit(n *parser.Select, p planNode) (planNode, error) {
+	if n.Limit == nil {
+		return p, nil
+	}
+
+	var count, offset int64
+
+	data := []struct {
+		name       string
+		src        parser.Expr
+		dst        *int64
+		defaultVal int64
+	}{
+		{"LIMIT", n.Limit.Count, &count, math.MaxInt64},
+		{"OFFSET", n.Limit.Offset, &offset, 0},
+	}
+
+	for _, datum := range data {
+		if datum.src == nil {
+			*datum.dst = datum.defaultVal
+		} else {
+			if parser.ContainsVars(datum.src) {
+				return nil, fmt.Errorf("argument of %s must not contain variables", datum.name)
+			}
+
+			normalized, err := parser.NormalizeExpr(datum.src)
+			if err != nil {
+				return nil, err
+			}
+			dstDatum, err := parser.EvalExpr(normalized)
+			if err != nil {
+				return nil, err
+			}
+
+			if dstDatum == parser.DNull {
+				*datum.dst = datum.defaultVal
+				continue
+			}
+
+			if dstDInt, ok := dstDatum.(parser.DInt); ok {
+				*datum.dst = int64(dstDInt)
+				continue
+			}
+
+			return nil, fmt.Errorf("argument of %s must be type %s, not type %s", datum.name, parser.DummyInt.Type(), dstDatum.Type())
+		}
+	}
+
+	return &limitNode{planNode: p, count: count, offset: offset}, nil
+}
+
+type limitNode struct {
+	planNode
+	count          int64
+	offset         int64
+	rowIndex       int64
+	outputRowIndex int64
+}
+
+func (n *limitNode) Next() bool {
+	if n.outputRowIndex >= n.count {
+		return false
+	}
+
+	for n.rowIndex < n.offset && n.planNode.Next() {
+		n.rowIndex++
+	}
+
+	n.outputRowIndex++
+	return n.planNode.Next()
+}
+
+func (n *limitNode) ExplainPlan() (string, string, []planNode) {
+	var limit string
+	if n.count == math.MaxInt64 {
+		limit = "NONE"
```

Comment:
Queries can specify `LIMIT ALL`. I think it would make more sense to `s/NONE/ALL/g` here to stay consistent with that terminology.